certs/mocks/things.go

```diff
@@ -0,0 +1,279 @@
+// Copyright (c) Mainflux
+// SPDX-License-Identifier: Apache-2.0
+
+package mocks
+
+import (
+	"context"
+	"strconv"
+	"sync"
+
+	"github.com/mainflux/mainflux"
+	"github.com/mainflux/mainflux/auth"
+	"github.com/mainflux/mainflux/things"
+)
+
+var _ things.Service = (*mainfluxThings)(nil)
+
+type mainfluxThings struct {
+	mu          sync.Mutex
+	counter     uint64
+	things      map[string]things.Thing
+	channels    map[string]things.Channel
+	auth        mainflux.AuthServiceClient
+	connections map[string][]string
+}
+
+// NewThingsService returns Mainflux Things service mock.
+// Only methods used by SDK are mocked.
+func NewThingsService(things map[string]things.Thing, channels map[string]things.Channel, auth mainflux.AuthServiceClient) things.Service {
+	return &mainfluxThings{
+		things:      things,
+		channels:    channels,
+		auth:        auth,
+		connections: make(map[string][]string),
+	}
+}
+
+func (svc *mainfluxThings) CreateThings(_ context.Context, owner string, ths ...things.Thing) ([]things.Thing, error) {
+	svc.mu.Lock()
+	defer svc.mu.Unlock()
+
+	userID, err := svc.auth.Identify(context.Background(), &mainflux.Token{Value: owner})
+	if err != nil {
+		return []things.Thing{}, things.ErrUnauthorizedAccess
+	}
+	for i := range ths {
+		svc.counter++
+		ths[i].Owner = userID.Email
+		ths[i].ID = strconv.FormatUint(svc.counter, 10)
+		ths[i].Key = ths[i].ID
+		svc.things[ths[i].ID] = ths[i]
+	}
+
+	return ths, nil
+}
+
+func (svc *mainfluxThings) ViewThing(_ context.Context, owner, id string) (things.Thing, error) {
+	svc.mu.Lock()
+	defer svc.mu.Unlock()
+
+	userID, err := svc.auth.Identify(context.Background(), &mainflux.Token{Value: owner})
+	if err != nil {
+		return things.Thing{}, things.ErrUnauthorizedAccess
+	}
+
+	if t, ok := svc.things[id]; ok && t.Owner == userID.Email {
+		return t, nil
+
+	}
+
+	return things.Thing{}, things.ErrNotFound
+}
+
+func (svc *mainfluxThings) Connect(_ context.Context, owner string, chIDs, thIDs []string) error {
+	svc.mu.Lock()
+	defer svc.mu.Unlock()
+
+	userID, err := svc.auth.Identify(context.Background(), &mainflux.Token{Value: owner})
+	if err != nil {
+		return things.ErrUnauthorizedAccess
+	}
+	for _, chID := range chIDs {
+		if svc.channels[chID].Owner != userID.Email {
+			return things.ErrUnauthorizedAccess
+		}
+		svc.connections[chID] = append(svc.connections[chID], thIDs...)
+	}
+
+	return nil
+}
+
+func (svc *mainfluxThings) Disconnect(_ context.Context, owner, chanID, thingID string) error {
+	svc.mu.Lock()
+	defer svc.mu.Unlock()
+
+	userID, err := svc.auth.Identify(context.Background(), &mainflux.Token{Value: owner})
+	if err != nil || svc.channels[chanID].Owner != userID.Email {
+		return things.ErrUnauthorizedAccess
+	}
+
+	ids := svc.connections[chanID]
+	i := 0
+	for _, t := range ids {
```

Comment:
What is this doing? Probably some comments are needed here.

Also, can this search be improved?

Reply:
this file `things.go` from other service is copied , there is opened issue for refactoring mocks so that we dont copy in each service same code so this could be part of that PR https://github.com/mainflux/mainflux/issues/1382